avm/res/network/virtual-network/main.bicep

```diff
@@ -130,12 +130,12 @@ resource virtualNetwork 'Microsoft.Network/virtualNetworks@2023-04-01' = {
             : []
           delegations: contains(subnet, 'delegations') ? subnet.delegations : []
           ipAllocations: contains(subnet, 'ipAllocations') ? subnet.ipAllocations : []
-          natGateway: contains(subnet, 'natGatewayResourceId')
+          natGateway: contains(subnet, 'natGatewayResourceId') && !empty(subnet.natGatewayResourceId)
             ? {
                 id: subnet.natGatewayResourceId
               }
             : null
-          networkSecurityGroup: contains(subnet, 'networkSecurityGroupResourceId')
+          networkSecurityGroup: contains(subnet, 'networkSecurityGroupResourceId') && !empty(subnet.networkSecurityGroupResourceId)
```

Comment:
I'm not sure if this is a great idea. I recall we used to have a related discussion about empty checks a long time ago in CARML, and the main catch was that as soon as you add this check, you're moving yourself into a corner.

To be precise, you can, for example, always set the networkSecurityGroup, but once you did, you can never unset it again (back to no network security group). Please correct me if I'm wrong.

Reply:
It does remove the NSG.

My first module call:
```bicep
module vnet '../avm/res/network/virtual-network/main.bicep' = {
  name: 'vnet-add'
  scope: rg
  params: {
    name: 'vnet-tests'
    location: rg.location
    addressPrefixes: ['10.0.0.0/24']
    subnets: [
      {
        name: 'default'
        addressPrefix: '10.0.0.0/26'
        networkSecurityGroupResourceId: nsg.outputs.resourceId
      }
    ]
  }
}
```
Result:
![image](https://github.com/Azure/bicep-registry-modules/assets/204027/a45de725-86f7-4271-928b-c55f1350ba92)


My second module call:
```bicep
module vnet '../avm/res/network/virtual-network/main.bicep' = {
  name: 'vnet-add'
  scope: rg
  params: {
    name: 'vnet-tests'
    location: rg.location
    addressPrefixes: ['10.0.0.0/24']
    subnets: [
      {
        name: 'default'
        addressPrefix: '10.0.0.0/26'
        networkSecurityGroupResourceId: ''
      }
    ]
  }
}
```
Result:
![image](https://github.com/Azure/bicep-registry-modules/assets/204027/2357c8b6-e078-409e-b4ae-477663f0fcb7)


But after your comments yesterday, I'm ultimately undecided whether this is the right way to go. 🤷‍♂️
Your solution using the union() function also works, even if I think it's not that "discoverable" or easy to use by end users.

Let me know what you think.

Reply:
Forgot the third test - Removing the parameter.

After re-applying the NSG with the first template, I've used this module call and it also has removed the NSG:
```bicep
module vnet '../avm/res/network/virtual-network/main.bicep' = {
  name: 'vnet-add'
  scope: rg
  params: {
    name: 'vnet-tests'
    location: rg.location
    addressPrefixes: ['10.0.0.0/24']
    subnets: [
      {
        name: 'default'
        addressPrefix: '10.0.0.0/26'
      }
    ]
  }
}
```

Reply:
Hey @cloudchristoph,
thanks for the due dilligence. That's good to see. In this case I'd pivot towards your suggestion as it may add some code to the template and is ultimately not needed when using the `union()`, yet from where I'm standing it's simply adding another case handling (i.e., somebody wants to provide an empty value) without taking anything away. 

What do you think?

Reply:
Yes, it doesn't take away any functionality, is backwards compatible and simplifies writing.

For now, there is nothing I would add. In the future maybe we could write tests that test both adding and removing route tables / NSGs etc.

Reply:
There certainly is an option. A test template could deploy one instance with and without the parameter and as we support post-deployment pester tests, we could validate that a subnet in question has a property set or not.